### PR TITLE
[Gamemode] Headhunter (again)

### PR DESCRIPTION
**Pull request by Coopyy**
"hey didnt he already make a pr for this"
well its been 8 months and theres still no thunderstore so im back (karma told me to)
https://github.com/R2Northstar/NorthstarMods/pull/187/

### Headhunter

kinda like the faster paced sibling to Fastball
works on all maps (yes live fire maps, just dont get thrown)

You're thrown into the map by BT and pilots are eliminated when killed
pilots drop a battery on death, you can pick up multiple batteries
for each battery you return, you revive a dead teammate
if time runs out team with more players alive wins

(also fixed crash with dying with more than 4 batteries :] )

medias:
great videos and gameplay from iniquity: 
https://youtu.be/Hqrz7y-RNbg
https://youtu.be/JweMsade_Ag
heres a Macro video
https://www.youtube.com/watch?v=Mf_-0vAM1LM

uh

uh2

### CODE REVIEW

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 1]
```diff
@@ -0,0 +1,291 @@
+untyped
+global function GamemodeHH_Init
+
+const array<string> btmaps = ["mp_angel_city", "mp_colony02", "mp_thaw", "mp_glitch", "mp_wargames", "mp_drydock", "mp_eden", "mp_black_water_canal", "mp_grave", "mp_complex3", "mp_homestead", "mp_forwardbase_kodai", "mp_crashsite3", "mp_rise", "mp_relic02"]
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:4]:
```suggestion
const array<string> btmaps = [ "mp_angel_city", "mp_colony02", "mp_thaw", "mp_glitch", "mp_wargames", "mp_drydock", "mp_eden", "mp_black_water_canal", "mp_grave", "mp_complex3", "mp_homestead", "mp_forwardbase_kodai", "mp_crashsite3", "mp_rise", "mp_relic02" ]
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+struct {
+	bool hasmadespawns = false
+	array<entity> goals
+	array<vector> spawnoffsets = [Vector(0,50,10), Vector(0,-50,10), Vector(50,0,10), Vector(-50,-0,10)] // some sort of spawn checking in the future
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:9]:
```suggestion
	array<vector> spawnoffsets = [ Vector(0,50,10), Vector(0,-50,10), Vector(50,0,10), Vector(-50,-0,10) ] // some sort of spawn checking in the future
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+} file
+
+void function GamemodeHH_Init()
+{
+	// used for intro
+	if (btmaps.contains(GetMapName()))
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:15]:
```suggestion
	if ( btmaps.contains( GetMapName() ) ) 
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+	
+	SetShouldUseRoundWinningKillReplay( true )
+	SetRoundBased( true )
+	SetRespawnsEnabled( false )
+	Riff_ForceTitanAvailability( eTitanAvailability.Never )
+	Riff_ForceSetEliminationMode( eEliminationMode.Pilots )
+	ScoreEvent_SetupEarnMeterValuesForMixedModes()
+
+	ClassicMP_ForceDisableEpilogue( true )
+	PrecacheModel($"models/robots/mobile_hardpoint/mobile_hardpoint.mdl")
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:34]:
```suggestion
	PrecacheModel( $"models/robots/mobile_hardpoint/mobile_hardpoint.mdl" )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+	PilotBattery_SetMaxCount( 99 )
+}
+
+void function StartHHThink() {
+	thread HHThink()
+}
+
+void function HHThink() {
+	while(true)
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:52]:
```suggestion
	while( true ) 
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+void function StartHHThink() {
+	thread HHThink()
+}
+
+void function HHThink() {
+	while(true) 
+	{
+		foreach (entity player in GetPlayerArray())
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:54]:
```suggestion
		foreach ( entity player in GetPlayerArray() )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+		{
+			if (!PlayerHasBattery(player))
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:56]:
```suggestion
			if ( !PlayerHasBattery( player ) )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+				continue
+			foreach (entity goal in file.goals)
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:58]:
```suggestion
			foreach ( entity goal in file.goals ) 
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+void function HHThink() {
+	while(true) 
+	{
+		foreach (entity player in GetPlayerArray())
+		{
+			if (!PlayerHasBattery(player))
+				continue
+			foreach (entity goal in file.goals) 
+			{
+				if (DistanceSqr(player.GetOrigin(), goal.GetOrigin()) <= 30000.0)
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:60]:
```suggestion
				if ( DistanceSqr( player.GetOrigin(), goal.GetOrigin() ) <= 30000.0 ) 
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+			if (!PlayerHasBattery(player))
+				continue
+			foreach (entity goal in file.goals) 
+			{
+				if (DistanceSqr(player.GetOrigin(), goal.GetOrigin()) <= 30000.0) 
+				{
+					array<entity> possPlayers
+					foreach ( entity deadPlayer in GetPlayerArrayOfTeam( player.GetTeam() ) )
+						if ( !IsAlive( deadPlayer ) )
+							possPlayers.append(deadPlayer)
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:65]:
```suggestion
							possPlayers.append( deadPlayer )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+					if (possPlayers.len() > 0)
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:66]:
```suggestion
					if ( possPlayers.len() ) 
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+			{
+				if (DistanceSqr(player.GetOrigin(), goal.GetOrigin()) <= 30000.0) 
+				{
+					array<entity> possPlayers
+					foreach ( entity deadPlayer in GetPlayerArrayOfTeam( player.GetTeam() ) )
+						if ( !IsAlive( deadPlayer ) )
+							possPlayers.append(deadPlayer)
+					if (possPlayers.len() > 0) 
+					{
+						entity chosen = possPlayers[RandomInt(possPlayers.len())]
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:68]:
```suggestion
						entity chosen = possPlayers[ RandomInt( possPlayers.len() ) ]
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+						if (!IsAlive(chosen)) {
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:69]:
```suggestion
						if ( !IsAlive( chosen ) )
						{
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+							chosen.SetOrigin(goal.GetOrigin() + file.spawnoffsets[RandomInt(file.spawnoffsets.len())])
+							chosen.RespawnPlayer( null )
+							Remote_CallFunction_NonReplay(chosen, "ServerCallback_HeadhunterRespawnPlayer", player.GetEncodedEHandle())
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:72]:
```suggestion
							chosen.SetOrigin( goal.GetOrigin() + file.spawnoffsets[ RandomInt( file.spawnoffsets.len() ) ] )
							chosen.RespawnPlayer( null )
							Remote_CallFunction_NonReplay( chosen, "ServerCallback_HeadhunterRespawnPlayer", player.GetEncodedEHandle() )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+					}
+				}
+			} 
+		}
+
+		wait 0.1
+	}
+}
+
+void function RemoveBatts() {
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:87]:
```suggestion
void function RemoveBatts()
{
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+	foreach (entity player in GetPlayerArray())
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:88]:
```suggestion
	foreach ( entity player in GetPlayerArray() )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+void function SetClientRui( entity player )
+{
+	thread SetClientRuiThread(player)
+}
+
+void function SetClientRuiThread(entity player) 
+{
+	while (!(file.goals.len() > 2))
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:122]:
```suggestion
	while ( !( file.goals.len() > 2 ) )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+		wait 0.5
+
+	if (!IsValid(player))
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:125]:
```suggestion
	if ( !IsValid( player ) )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+		return
+	
+	int k = 0
+	while (!(k > 2)) {
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:129]:
```suggestion
	while ( !( k > 2 ) )
	{
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+		for (int i = 0; i < file.goals.len(); i++)
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:130]:
```suggestion
		for ( int i; i < file.goals.len(); i++ ) 
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+		wait 0.5
+
+	if (!IsValid(player))
+		return
+	
+	int k = 0
+	while (!(k > 2)) {
+		for (int i = 0; i < file.goals.len(); i++) 
+		{
+			if (!PlayerInRenderDistance(file.goals[i])) // why the FUCK does this work
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:132]:
```suggestion
			if ( !PlayerInRenderDistance( file.goals[i] ) ) // why the FUCK does this work
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+				continue
+			Remote_CallFunction_NonReplay( player, "ServerCallback_HeadhunterUpdateRui", file.goals[i].GetEncodedEHandle(), i )
+			k++
+		}
+		wait 0.1
+	}
+}
+
+bool function PlayerInRenderDistance(entity origin)
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:141]:
```suggestion
bool function PlayerInRenderDistance( entity origin )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+			Remote_CallFunction_NonReplay( player, "ServerCallback_HeadhunterUpdateRui", file.goals[i].GetEncodedEHandle(), i )
+			k++
+		}
+		wait 0.1
+	}
+}
+
+bool function PlayerInRenderDistance(entity origin) 
+{
+	if (!IsValid(origin))
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:143]:
```suggestion
	if ( !IsValid( origin ) )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 24]
```diff
@@ -0,0 +1,291 @@
...
+		return false
+	foreach (entity player in GetPlayerArray())
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:145]:
```suggestion
	foreach ( entity player in GetPlayerArray() )
```

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+	}
+}
+
+bool function PlayerInRenderDistance(entity origin) 
+{
+	if (!IsValid(origin))
+		return false
+	foreach (entity player in GetPlayerArray()) 
+	{
+		if (Distance(player.GetOrigin(), origin.GetOrigin()) < 7500)
```
Comment (uniboi) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:147]:
```suggestion
		if ( Distance( player.GetOrigin(), origin.GetOrigin() ) < 7500 ) 
```

**Review by uniboi — APPROVED** [review]
there's still a lot of formatting issues

**Review by uniboi — CHANGES_REQUESTED** [review]
github truly has one of the shittiest review systems

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut (hunk 2 above)
Comment (BobTheBob9) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:9]: why not use the vector literal here rather than a tf1-style constructor?

**Review by BobTheBob9 — CHANGES_REQUESTED** [review]
needs testing and localization before merge, but looks pretty good

**Comment by Coopyy** [comment]
why must there be 25 separate suggestions i must commit

that is 15 minutes of my life vanished

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut (hunk 2 above)
Comment (Coopyy) [reply Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut]: uh uh uh idk

**Comment by GeckoEidechse** [comment]
> why must there be 25 separate suggestions i must commit
> 
> that is 15 minutes of my life vanished

FYI, if you go to the _Files changed_ tab, you can queue up accepting suggested changes and commit them all at once. GitHub just doesn't make this obvious at all ^^

**Comment by Coopyy** [comment]
what is it whining about

File: Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut [hunk 26]
```diff
@@ -0,0 +1,294 @@
+untyped
+global function GamemodeHH_Init
+
+const array<string> btmaps = [ "mp_angel_city", "mp_colony02", "mp_thaw", "mp_glitch", "mp_wargames", "mp_drydock", "mp_eden", "mp_black_water_canal", "mp_grave", "mp_complex3", "mp_homestead", "mp_forwardbase_kodai", "mp_crashsite3", "mp_rise", "mp_relic02" ]
+
+struct {
+	bool hasmadespawns = false
+	array<entity> goals
+	array<vector> spawnoffsets = [ <0,50,10>, <0,-50,10>, <50,0,10>, <-50,-0,10> ] // some sort of spawn checking (ray trace?) in the future
```
Comment (ASpoonPlaysGames) [inline Northstar.Custom/mod/scripts/vscripts/gamemodes/_gamemode_hh.gnut:9]:
```suggestion
	array<vector> spawnoffsets = [ < 0, 50, 10 >, < 0, -50, 10 >, < 50, 0, 10 >, < -50, -0, 10 > ] // some sort of spawn checking (ray trace?) in the future
```
its whining about this (no spaces meant it thought `<-50` was the arrow operator)

**Comment by Coopyy** [comment]
ebic

**Comment by GeckoEidechse** [comment]
With progress on auto-mod-download, I think this is one of the gamemodes that should get verified via [VerifiedMods](https://github.com/R2Northstar/VerifiedMods) repo so that players can auto-download it on server join :D

**Comment by GeckoEidechse** [comment]
And just to be clear mods get verified way faster than PRs get merged as verification only checks for malicious code, we don't really check for correctness etc ^^

**Comment by GeckoEidechse** [comment]
bump

**Comment by GeckoEidechse** [comment]
@Coopyy I reuploaded the PR as a standalone mod now via https://github.com/GeckoEidechse/Headhunter/ and made a verification request via https://github.com/R2Northstar/VerifiedMods/pull/42

If you'd rather have the mod be published entirely under your own name, please LMK and get in touch and I will retract my verification request and Thunderstore upload and you can publish it under your own name and request verification from there.

Either way, I'd say it's best to close this PR now as the gamemode makes more sense to be released standalone and downloaded automatically on join using mod-auto-download (MAD) as soon as https://github.com/R2Northstar/Northstar/issues/674 is complete

**Pull request closed without merging**